5000user5000 proposal for mpGEMM

5000user5000 opened the pull request:
<!-- Please fill the following checkboxes -->
---
A good PR should include the following parts:

- [X] Name your branch like ``<github-user-name>-hw#``, e.g., ``myname-hw2-attempt1``
- [X] Provide a clear title
- [X] Provide meaningful descriptions in the PR and the commit logs
- [X] Merge from your specific branch (***not master***) to the `hw*` branch (***not master***, again)
- [X] Exclude irrelevant files, such as binaries, text files, or dot files
- [X] Make sure tests/CI pass

yungyuc commented on project/5000user5000/README.rst:
```diff
@@ -0,0 +1,87 @@
+- **Project**: mpGEMM
+- **Github**: https://github.com/5000user5000/mpGEMM
+
+On resource-constrained devices, such as embedded systems, running deep learning models, especially LLMs, is highly computationally expensive. Low-bit quantization is a popular solution to reduce memory consumption. However, when weights are quantized below 8 bits, performing matrix multiplication between the weight matrix and an FP16 activation matrix requires dequantization to a common precisio ...
```
Please use 79-char line width.  Long lines (or single line per paragraph) make it hard to comment for review.

yungyuc commented on project/5000user5000/README.rst:
```diff
@@ -0,0 +1,87 @@
+Engineering Infrastructure
+==========================
+
+- **Automated Build System:** Uses CMake to set up the C++ build system and setuptools to build Python packages.
```
Where, how, and when (in schedule) to test?

yungyuc commented on project/5000user5000/README.rst:
```diff
@@ -0,0 +1,87 @@
+Problem to Solve
+================
+
+- Develop a **low-level numerical computation core** in C++ for high-speed matrix operations.
```
You won't have a lot to present if your implementation is slower than vendor-provided libraries (MKL, Accelerate, etc.), but it is very hard to beat the performance of vendor library.

Do you have an idea how much faster vendor library is compared to naive implementation?

5000user5000 commented:
I have adjusted the text to a 79-character line width and added a detailed test plan to ensure comprehensive testing.
Regarding the Problem to Solve section, I have removed the part about competing with vendor libraries, as it is indeed challenging to surpass their performance.

yungyuc reviewed (changes requested):
> I have adjusted the text to a 79-character line width and added a detailed test plan to ensure comprehensive testing. Regarding the Problem to Solve section, I have removed the part about competing with vendor libraries, as it is indeed challenging to surpass their performance.

You did not answer my question about whether or not you know how fast vendor libraries run.  Please include that in your proposal.  You do not need to be accurate.  Simple estimate is good enough.

To make a presentation for array operation interesting, speed is a must.

yungyuc commented:
Leave a global comment when you are ready for the next review.

- [x] Follow [`proposal/README.rst`](https://github.com/yungyuc/nsdhw_25sp/blob/f83e66ee5f60a614fbe5614581d5d09ced2c344d/project/README.rst?plain=1#L73) to add your proposal in it.
- [x] Use dates in your schedule.  Make testing obvious.

yungyuc commented:
@5000user5000 please also take a look at the in-progress work to develop a simple feature of columnar array container: https://github.com/solvcon/modmesh/pull/469, to get a taste for how much work it may take for designing interface and implementing the calculation kernel.

5000user5000 commented:
I have updated my proposal.
- Revised the schedule to include specific dates.
- Included an estimate of vendor library performance (10× to 1000× faster than naive GEMM).
- Removed the plan to implement a custom computation kernel and instead focused on matrix multiplication with lookup tables.

yungyuc commented on project/5000user5000/README.rst:
```diff
@@ -0,0 +1,154 @@
+processes submitted to the GitHub repository.
+
+
+Schedule
```
Add your plan of testing in the schedule.

yungyuc commented on project/5000user5000/README.rst:
```diff
@@ -0,0 +1,154 @@
+
+- Enable the LUT to reside in the fastest on-chip memory and parallel lookup.
+
+- Preliminary estimates suggest that vendor libraries such as MKL or
```
It's clear.  Let's see how it goes.

yungyuc commented on project/5000user5000/README.rst:
```diff
@@ -0,0 +1,154 @@
+  gemm = mpGEMM(backend="lut", use_simd=True)
+
+  # Perform GEMM operation, specifying weight bit-width
+  output = gemm.matmul(weights, activations, weight_bit_width=4)
```
Focusing on only matmul will make your Python interface too boring to make presentations.  Keep it in mind and see how it goes.

yungyuc reviewed (changes requested):
- [x] Add testing plan in schedule.

5000user5000 commented:
I have updated my proposal.
- Add plan of testing in the schedule.
- Enhanced Python interface with additional functions to make it more engaging.

yungyuc reviewed (approved):
LGTM.  Let me know when you want it to be merged.

5000user5000 commented:
I'm ready to merge. Thanks!
